src/thor/timedistancematrix.cc

```diff
@@ -84,33 +97,70 @@ void TimeDistanceMatrix::ExpandForward(GraphReader& graphreader,
     return;
   }
 
+  const DirectedEdge* opp_pred_edge = nullptr;
+  if (!FORWARD) {
+    opp_pred_edge = tile->directededge(nodeinfo->edge_index());
+    for (uint32_t i = 0; i < nodeinfo->edge_count(); i++, opp_pred_edge++) {
+      if (opp_pred_edge->localedgeidx() == pred.opp_local_idx()) {
+        break;
+      }
+    }
+  }
+
   // Expand from end node.
   GraphId edgeid(node.tileid(), node.level(), nodeinfo->edge_index());
   EdgeStatusInfo* es = edgestatus_.GetPtr(edgeid, tile);
   const DirectedEdge* directededge = tile->directededge(nodeinfo->edge_index());
   for (uint32_t i = 0; i < nodeinfo->edge_count(); i++, directededge++, ++edgeid, ++es) {
     // Skip shortcut edges
-    if (directededge->is_shortcut()) {
+    if (directededge->is_shortcut() || es->set() == EdgeSet::kPermanent) {
       continue;
     }
 
+    graph_tile_ptr t2 = nullptr;
+    GraphId opp_edge_id;
+    const DirectedEdge* opp_edge = nullptr;
+    if (!FORWARD) {
+      // Get opposing edge Id and end node tile
+      graph_tile_ptr t2 =
+          directededge->leaves_tile() ? graphreader.GetGraphTile(directededge->endnode()) : tile;
+      if (t2 == nullptr) {
+        continue;
+      }
+      GraphId opp_edge_id = t2->GetOpposingEdgeId(directededge);
+      opp_edge = t2->directededge(opp_edge_id);
+    }
+
     // Skip this edge if permanently labeled (best path already found to this
     // directed edge), if no access is allowed to this edge (based on costing
     // method), or if a complex restriction prevents this path.
     uint8_t restriction_idx = -1;
     const bool is_dest = dest_edges_.find(edgeid) != dest_edges_.cend();
-    if (es->set() == EdgeSet::kPermanent ||
-        !costing_->Allowed(directededge, is_dest, pred, tile, edgeid, 0, 0, restriction_idx) ||
-        costing_->Restricted(directededge, pred, edgelabels_, tile, edgeid, true)) {
-      continue;
+    if (FORWARD) {
+      if (!costing_->Allowed(directededge, is_dest, pred, tile, edgeid, 0, 0, restriction_idx) ||
+          costing_->Restricted(directededge, pred, edgelabels_, tile, edgeid, true)) {
+        continue;
+      }
+    } else {
+      // TODO: is the Restricted() function correct here passing directed_edge & pred & tile?
+      if (!costing_->AllowedReverse(directededge, pred, opp_edge, t2, opp_edge_id, 0, 0,
+                                    restriction_idx) ||
+          (costing_->Restricted(directededge, pred, edgelabels_, tile, edgeid, FORWARD))) {
```

Comment:
I copied most things from the `ExpandReverse` of course, but I do wonder if this is really correct.. Shouldn't we pass the `opp_edge` and `opp_pred_edge` and `t2` here?

Reply:
never mind.. all other reverse trees also work with the forward edge, so I guess this is correct..